### PR TITLE
refactor: Loadable mixin no longer declares onMount and onRemove

### LINKED ISSUE #1210: Remove onMount/onRemove from Loadable
## What could be improved

The `Loadable` mixin is required for the generic `Game` class; at the same time methods `onMount`/`onRemove` are only suitable for the `Component` class and the derived `FlameGame`.

These methods should, therefore, be moved out of `Loadable` and into the `Component` class.

- [x] Yes, I can work on a PR for this

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

`Loadable` mixin is currently used for 2 classes: `Game` and `Component`. Its purpose is to provide the caching layer for the future returned by the `onLoad()` method. Having empty methods `onMount()` / `onRemove()` in that mixin does not fit its overall purpose. (Plus, in general, the purpose of any mixin is to add some additional **concrete** functionality into an existing class).

In addition to the scope mismatch, these methods perform similar yet not quite the same function in `Game` and `Component`. Having these methods declared directly within `Game` and `Component` allows to document them more precisely, and help the user understand their functionality/purpose better.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

Closes #1210

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Comment by st-pasha** [comment]
Note: I wasn't quite able to figure out the difference between `onMount` and `onAttach` for the `Game` class -- which I guess only stresses the need for clear and precise documentation for that function.

**Comment by spydon** [comment]
> Note: I wasn't quite able to figure out the difference between `onMount` and `onAttach` for the `Game` class -- which I guess only stresses the need for clear and precise documentation for that function.

`onAttach` is when the Flutter widget is attached to the Flutter widget tree, `onMount` is when the component is added to a component tree (after `onLoad`).

File: packages/flame_bloc/test/flame_bloc_game_test.dart [hunk 1]
```diff
@@ -96,6 +96,7 @@ void main() {
         game.add(component);
 
         expect(component.state, equals(InventoryState.sword));
+        game.remove(component);
```
Comment (spydon) [inline packages/flame_bloc/test/flame_bloc_game_test.dart:99]: Why was this line added?

**Review by spydon — APPROVED** [review]
Lgtm

File: packages/flame_bloc/test/flame_bloc_game_test.dart (hunk 1 above)
Comment (st-pasha) [reply packages/flame_bloc/test/flame_bloc_game_test.dart:99]: In order to get 100% coverage for flame_bloc repository.

**Comment by st-pasha** [comment]
> onAttach is when the Flutter widget is attached to the Flutter widget tree, onMount is when the component is added to a component tree (after onLoad).

Exactly: this is where it gets confusing. I understand `onMount` for the `Component` class. But the Game class is a different story. So the question is: what is `onMount` in the `Game` class, and how is it different from the `onAttach` handler?

The situation with `onRemove` / `onDetach` is similar, only I believe the former is called _before_ the game is removed from the widget, while the latter called _after_. We should consider at some point to rename them to improve clarity. For example: `onBeforeDetach` / `onAfterDetach`.

**Comment by spydon** [comment]
> Exactly: this is where it gets confusing. I understand `onMount` for the `Component` class. But the Game class is a different story. So the question is: what is `onMount` in the `Game` class, and how is it different from the `onAttach` handler?

Yeah, I see the confusion. We should remove `onMount` and `onRemove` from the `game` mixin (I must have accidentally put them there) and in `FlameGame` we should call `onMount` from `onAttach` and `onRemove` from `onDetach` probably. Since a game can be added as a child to another component, there will still be a difference to them.

File: packages/flame_bloc/test/flame_bloc_game_test.dart (hunk 1 above)
Comment (spydon) [reply packages/flame_bloc/test/flame_bloc_game_test.dart:99]: But it is kind of cheating, this test doesn't test anything regarding removal?

Comment (st-pasha) [reply packages/flame_bloc/test/flame_bloc_game_test.dart:110]: Fair, i've expanded this to a bigger test

File: packages/flame/lib/src/game/mixins/game.dart [hunk 2]
```diff
@@ -95,6 +95,8 @@ mixin Game on Loadable {
   /// Use for calculating the FPS.
   void onTimingsCallback(List<FrameTiming> timings) {}
 
+  void onMount() {}
```
Comment (spydon) [inline packages/flame/lib/src/game/mixins/game.dart:98]: I don't think `onMount` and `onRemove` is needed in this mixin since it is meant for games that live can live in the component tree, for other games `onAttach` and `onDetach` can be used.

Comment (st-pasha) [reply packages/flame/lib/src/game/mixins/game.dart:98]: Ok, great, I'll remove them then.

Comment (st-pasha) [reply packages/flame/lib/src/game/mixins/game.dart:98]: I just tried removing them, and it turns out to be a breaking change -- even 2 tests start to fail. 
So, probably such breaking change should be in its own PR.
